File: server/src/main/java/com/fluffy/submission/application/SubmissionService.java
package com.fluffy.submission.application;

import com.fluffy.auth.domain.Member;
import com.fluffy.auth.domain.MemberRepository;
import com.fluffy.exam.domain.Exam;
import com.fluffy.exam.domain.ExamRepository;
import com.fluffy.global.exception.BadRequestException;
import com.fluffy.submission.application.request.SubmissionAppRequest;
import com.fluffy.submission.domain.Submission;
import com.fluffy.submission.domain.SubmissionRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class SubmissionService {

    private final ExamRepository examRepository;
    private final SubmissionRepository submissionRepository;
    private final MemberRepository memberRepository;
    private final SubmissionMapper submissionMapper;
    private final SubmissionLockService submissionLockService;

    @Transactional
    public void submit(SubmissionAppRequest request) {
        Member member = memberRepository.findByIdOrThrow(request.accessor().id());
        Exam exam = examRepository.findByIdOrThrow(request.examId());

        if (exam.isNotPublished()) {
            throw new BadRequestException("시험이 공개되지 않았습니다.");
        }

        if (exam.isSingleAttempt()) {
            String lockName = "submit:%d:%d".formatted(exam.getId(), member.getId());
            submissionLockService.submitWithLock(request, exam, member, lockName);
            return;
        }

        Submission submission = submissionMapper.toSubmission(exam, member.getId(), request);
        submissionRepository.save(submission);
    }
}

File: server/src/main/kotlin/com/fluffy/submission/application/SubmissionService.kt
package com.fluffy.submission.application

import com.fluffy.exam.domain.ExamRepository
import com.fluffy.exam.domain.findByIdOrThrow
import com.fluffy.global.exception.BadRequestException
import com.fluffy.submission.application.request.SubmissionRequest
import com.fluffy.submission.domain.Submission
import com.fluffy.submission.domain.SubmissionRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class SubmissionService(
    private val examRepository: ExamRepository,
    private val submissionRepository: SubmissionRepository,
    private val submissionLockService: SubmissionLockService,
) {

    @Transactional
    fun submit(request: SubmissionRequest) {
        val memberId = request.accessor.id
        val exam = examRepository.findByIdOrThrow(request.examId)

        if (exam.isNotPublished()) {
            throw BadRequestException("시험이 공개되지 않았습니다.")
        }

        if (exam.isSingleAttempt) {
            val lockName = "submit:${exam.id}:${memberId}"
            submissionLockService.submitWithLock(request, exam, lockName)
            return
        }

        val submission: Submission = SubmissionAssembler.toSubmission(exam, memberId, request)
        submissionRepository.save(submission)
    }
}